control/src/http/attributes_builder.cc

```diff
@@ -0,0 +1,134 @@
+      {CheckData::HEADER_USER_AGENT, AttributeName::kRequestUserAgent, false,
+       ""},
+  };
+
+  for (const auto& it : attrs) {
+    std::string data;
+    if (check_data->FindRequestHeader(it.header_type, &data)) {
+      builder.AddString(it.name, data);
+    } else if (it.set_default) {
+      builder.AddString(it.name, it.default_value);
+    }
+  }
+}
+
+void AttributesBuilder::ExtractForwardedAttributes(CheckData* check_data) {
+  std::string forwarded_data;
+  if (!check_data->ExtractIstioAttributes(&forwarded_data)) {
+    return;
+  }
+  Attributes v2_format;
+  if (v2_format.ParseFromString(forwarded_data)) {
+    request_->attributes.MergeFrom(v2_format);
+    return;
+  }
+
+  // Legacy format from old proxy.
+  Attributes_StringMap forwarded_attributes;
+  if (forwarded_attributes.ParseFromString(forwarded_data)) {
+    ::istio::mixer_client::AttributesBuilder builder(&request_->attributes);
+    for (const auto& it : forwarded_attributes.entries()) {
+      builder.AddIpOrString(it.first, it.second);
+    }
+  }
+}
+
+void AttributesBuilder::ExtractCheckAttributes(CheckData* check_data) {
+  ExtractRequestHeaderAttributes(check_data);
+
+  ::istio::mixer_client::AttributesBuilder builder(&request_->attributes);
+
+  std::string source_ip;
+  int source_port;
+  if (check_data->GetSourceIpPort(&source_ip, &source_port)) {
+    builder.AddBytes(AttributeName::kSourceIp, source_ip);
+    builder.AddInt64(AttributeName::kSourcePort, source_port);
+  }
+
+  std::string source_user;
+  if (check_data->GetSourceUser(&source_user)) {
+    builder.AddString(AttributeName::kSourceUser, source_user);
+  }
+  builder.AddTimestamp(AttributeName::kRequestTime,
+                       std::chrono::system_clock::now());
+  builder.AddString(AttributeName::kContextProtocol, "http");
+}
+
+void AttributesBuilder::ForwardAttributes(const Attributes& forward_attributes,
+                                          CheckData* check_data) {
+  std::string str;
+  forward_attributes.SerializeToString(&str);
+  check_data->AddIstioAttributes(str);
+}
+
+void AttributesBuilder::ExtractReportAttributes(ReportData* report_data) {
+  ::istio::mixer_client::AttributesBuilder builder(&request_->attributes);
+  std::map<std::string, std::string> headers =
+      report_data->GetResponseHeaders();
+  builder.AddStringMap(AttributeName::kResponseHeaders, headers);
+
+  builder.AddTimestamp(AttributeName::kResponseTime,
+                       std::chrono::system_clock::now());
+
+  ReportData::ReportInfo info;
+  report_data->GetReportInfo(&info);
+  builder.AddInt64(AttributeName::kRequestSize, info.received_bytes);
+  builder.AddInt64(AttributeName::kResponseSize, info.send_bytes);
+  builder.AddDuration(AttributeName::kResponseDuration, info.duration);
+  if (!request_->check_status.ok()) {
+    builder.AddInt64(
+        AttributeName::kResponseCode,
+        utils::StatusHttpCode(request_->check_status.error_code()));
```

Comment:
This conflates Mixer check error code and the user's response HTTP status code. Should we use a different approach (attribute?) to correlate Check status with Report?

/cc @douglas-reid